### DISCUSSION TITLE
how to cluster irregularly spaced time series?

### DISCUSSION THREAD

**Discussion by fkiraly** [alan-turing-institute/sktime#3904] [Q&A]
question from discord helpdesk:

Could anyone direct me to ways of using sktime for clustering if one are dealing with irregularly sampled multivariate time-series. These time-series are only about 30 samples long, and stems from hospital vitals registrations on pulse, temperature, blood pressure, etc. There could be long periods with no registrations, and then periods with high frequency registrations. Moreover, the time series could be of various lengths.

**Comment by fkiraly** [comment]
Generally, you want to use the [`all_estimators`](https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.registry.all_estimators.html) search utility to search for either clusterers that can deal with irregularly sampled time series, or transformers that turn irregularly sampled time series into regularly sampled ones.
You should be able to pipeline transformers with clusterers that can't deal with irregularly sampled ones to get a clusterer that can deal with irregularly sampled time series

Caveat: the clustering module is still fairly experimental.

The tags to search for:

* `"capability:unequal_length"` for clusterers
* `"capability:unequal_length:removes"` for transformers

Generally, you can use the `all_tags` utility to search for tags and their explanations.

**Reply by helgeingvart** [reply]
Regrettably, there seem to be no clusterer having the capability of unequal_length. Is there somewhere I can report this to a "wish-list" for christmas? :-)

**Reply by fkiraly** [reply]
I can write you a clustering pipeline, @helgeingvart, that way you could use the transformers to build clusterers capable of unequal length?
🎄 🎅 🎁 ...

**Reply by fkiraly** [reply]
🎁 https://github.com/sktime/sktime/pull/3967

**Reply by fkiraly** [reply]
🎁 https://github.com/sktime/sktime/pull/4003 - DBSCAN where it is possible to plug in distances capable of unequal length series.

**Reply by helgeingvart** [reply]
Wow, this needs to be tested! I've been playing around a bit lately with another approach called Reservoir Computing, there is an excellent notebook on the work from Filippo Bianchi here [https://github.com/FilippoMB/Time-series-classification-and-clustering-with-Reservoir-Computing/blob/master/code/clustering_example.ipynb] if you would like to have a look. 
However, since this framework didn't have the requested support for irregular sampling, I had to write an interpolation "middleware" for adaption.

**Reply by fkiraly** [reply]
@helgeingvart, nice!

Would be great if you write an `sktime` clusterer wrapper anyway for this, if you would consider contributing it to `sktime`! Should be relatively easy, here's a guide which also has a walkthrough video: https://www.sktime.org/en/stable/developer_guide/add_estimators.html
Your task is (obviously) clustering.

**Reply by fkiraly** [reply]
PS: the above should be in 0.15.1 which will be released any day now

**Reply by helgeingvart** [reply]
Sorry, didn't see this one until now :-) I have a student that is considering your latest changes #4003 and #3967. I may try to add this functionality to sktime, but regrettably I didn't get many promising results from it so far, but it may relate to poor quality of my data.

**Reply by fkiraly** [reply]
ah well, you can't always win, I suppose.
If you are up to sharing the data in one of the community sessions, some of the more active community members could have a look. @JonathanBechtel and team are working on clustering and usually attend the Friday meet-ups, so might be interested.

Also, if you have anything for a wishlist, please let us know (e.g., via a wishlist issue!)

**Reply by Escape142** [reply]
Hi @helgeingvart,

Just a quick question: has the functionality for irregularly sampled time series clustering been added to sktime?

Thanks!

**Reply by fkiraly** [reply]
as far as I know, no - none of the clusterers support passing of an arbitrary `sktime` distance, which would allow to construct clusterers ofr unequal length or irregularly sampled time series.

I'm happy to instruct on a mini-project if you would be willing to contribute and/or know someone who would. I estimate 10-20 hours for upgrading one or a few of the existing clusterers, or adding a new one based on an `sklearn` clusterer.

**Reply by Escape142** [reply]
What approach do you have in mind to suggest?
